docs/Controllers.md

```diff
@@ -0,0 +1,99 @@
+However, where the RAML specification of an API specifies explicit response codes the Controller SHOULD handle these cases explicitly.
+
+For Controllers performing `GET` and `HEAD` requests, using these methods SHOULD correctly handle a `301` (Moved Permanently) response.
+
+When a `301` is supported, the Controller MUST follow the redirect in order to retrieve the required response payload.
+
+If a Controller receives an HTTP `5xx` or `4xx` response code from the API, a failure has occurred.
+The Controller SHOULD display the content of the response's error field to the user if possible, and indicate that the resource is likely to be in an unknown state.
+The Controller SHOULD also refresh the endpoints of the relevant resources to ensure the Controller is accurately reflecting the current state of the API.
+
+## Registry Service Discovery	
+
+In order to locate the Query API, the Controller SHOULD support all of the following:
+* Unicast DNS Service Discovery (DNS-SD).
+* Multicast DNS Service Discovery (mDNS).
+* Direct configuration with the location of a preferred Query API.
+
+The Controller SHOULD offer unicast DNS-SD as the default mechanism. 
+
+Controllers SHOULD observe and interpret all of the TXT records returned with the DNS Service Discovery responses according to the requirements for Query API Clients specified in the [API Paths section of the APIs document](APIs.md#api-paths) in this specification.
+
+## Query API
+
+The Controller MUST be capable of using the Query API to discover any registered resource, including Node, Device, Source, Flow, Sender, and Receiver,
+as described in the [APIs section](README.md#apis) of this specification.
+
+The Controller MUST use the Query API either via the HTTP API or by requesting WebSocket subscriptions.
+
+When using the Query API, query filters SHOULD be used (and advanced query language where available) to cut down on the volume of resources returned to the Controller, as specified in the [APIs: Query Parameters](APIs%20-%20Query%20Parameters.md) document.
+
+Controllers SHOULD adhere to the version downgrade requirements for Query API Clients specified in the [Requirements for Query API Clients section of the Upgrade Path document](Upgrade%20Path.md#requirements-for-query-api-clients) in this specification.
+
+## Pagination
+
+In large systems, API resources rely upon pagination in order to return high volumes of data.
+This can be both hard to keep track of for a client, and require a large number of requests in order to scan the entire data set.
+
+For this reason it is RECOMMENDED to use the RESTful QUERY API for debug and development purposes only.
+
+If using the HTTP API rather than WebSocket subscriptions, pagination requirements MUST be implemented as specified in the [Pagination section of the APIs: Query Parameters document](APIs%20-%20Query%20Parameters.md#pagination) in this specification.
```

Comment:
```suggestion
If using the HTTP API rather than WebSocket subscriptions, pagination requirements MUST be implemented as specified in the [APIs: Query Parameters: Pagination](APIs%20-%20Query%20Parameters.md#pagination) section of this specification.
```